daemon/exec.go

```diff
@@ -274,31 +273,39 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
 	ec.StreamConfig.AttachStreams(&attachConfig)
 	attachErr := ec.StreamConfig.CopyStreams(ctx, &attachConfig)
 
+	ec.Container.Lock()
+	tsk, err := ec.Container.GetRunningTask()
+	ec.Container.Unlock()
+	if err != nil {
+		return err
+	}
+
 	// Synchronize with libcontainerd event loop
 	ec.Lock()
-	systemPid, err := daemon.containerd.Exec(ctx, c.ID, ec.ID, p, cStdin != nil, ec.InitializeStdio)
+	ec.Process, err = tsk.Exec(ctx, ec.ID, p, cStdin != nil, ec.InitializeStdio)
 	// the exec context should be ready, or error happened.
 	// close the chan to notify readiness
 	close(ec.Started)
 	if err != nil {
 		ec.Unlock()
 		return translateContainerdStartErr(ec.Entrypoint, ec.SetExitCode, err)
 	}
-	ec.Pid = systemPid
 	ec.Unlock()
 
 	select {
 	case <-ctx.Done():
-		logrus.Debugf("Sending TERM signal to process %v in container %v", name, c.ID)
-		daemon.containerd.SignalProcess(ctx, c.ID, name, signal.SignalMap["TERM"])
+		// Must use a new context since the current context is done.
+		ctx := context.TODO()
+		logrus.Debugf("Sending TERM signal to process %v in container %v", name, ec.Container.ID)
+		ec.Process.Kill(ctx, signal.SignalMap["TERM"])
```

Comment:
Relevant to #43739, this could be simpler if we were just `SIGKILL`ing :eyes: (this is the same bit of code, right?)

Reply:
It is the same bit of code!